packages/back-end/src/util/organization.util.ts

```diff
@@ -68,10 +71,77 @@ export function roleToPermissionMap(
   return permissionsObj;
 }
 
-export function getUserPermissions(
+function mergePermissions(
+  existingPermissions: UserPermission,
+  existingRole: MemberRole | undefined,
+  teamInfo: TeamInterface | ProjectMemberRole,
+  org: OrganizationInterface
+) {
+  const newPermissions = roleToPermissionMap(teamInfo.role, org);
+  for (const newPermission in newPermissions) {
+    // If the user doesn't have permission, but the team role does, add it
+    if (
+      !existingPermissions.permissions[newPermission as Permission] &&
+      newPermissions[newPermission as Permission]
+    ) {
+      existingPermissions.permissions[newPermission as Permission] =
+        newPermissions[newPermission as Permission];
+    }
+  }
+
+  if (!existingRole) {
+    existingPermissions.limitAccessByEnvironment =
+      teamInfo.limitAccessByEnvironment;
+    existingPermissions.environments = teamInfo.environments;
+  } else {
+    const roles = getRoles(org);
+    const newRoleIndex = roles.findIndex((r) => r.id === teamInfo.role);
+    const existingRoleIndex = roles.findIndex((r) => r.id === existingRole);
+
+    const existingRoleIsEngineerOrExperimenter =
+      existingRole === "engineer" || existingRole === "experimenter";
+
+    const newRoleIsEngineerOrExperimenter =
+      teamInfo.role === "engineer" || teamInfo.role === "experimenter";
+
+    if (
+      existingRoleIsEngineerOrExperimenter &&
+      newRoleIsEngineerOrExperimenter
+    ) {
+      if (
+        // If limitAccessByEnvironment is the same for new and existing roles, we just concat the envs arrays
+        existingPermissions.limitAccessByEnvironment ===
+        teamInfo.limitAccessByEnvironment
+      ) {
+        existingPermissions.environments = [
+          ...new Set(
+            existingPermissions.environments.concat(teamInfo.environments)
+          ),
+        ];
+      } else {
+        // otherwise, 1 role doesn't have limited access by environment, so it overrides the other
+        existingPermissions.limitAccessByEnvironment = false;
+        existingPermissions.environments = [];
+      }
+    } else {
+      // Finally, we set the limitAccessByEnvironment and environments properties to the more permissive role's values.
+      existingPermissions.limitAccessByEnvironment =
+        newRoleIndex > existingRoleIndex
```

Comment:
Why are we working with the index of the roles here? Are we making an assumption that the ordering of the roles indicates which is more permissive? If so, we may want to do this in a different way to avoid making that assumption since it could introduce some bugs if we don't adhere to that rule and to make this a bit more readable

Reply:
You're reading it correctly. The logic is a bit twisty - all of the roles, do get more permissive, with the exception of the `engineer` and `analyst` roles - those don't fit that pattern.

I agree that this could introduce bugs, though, I'm not sure of an alternative route.

I think this pattern can also make future updates for "custom" roles a bit more challenging as well.

I'll do some brainstorming on this (probably tomorrow AM when my mind is more fresh).